---
title: "Hw1_280"
author: "Sarah ji"
date: "1/18/2018"
output: html_document
---

```{r setup, include=FALSE}
#setwd("/Users/sarahji/Desktop/biostat-m280-2018-winter")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(evaluate = F) # on the teaching server this will be true since i have the files there
```

##Q1. Git/Github
1. Apply for the Student Developer Pack at GitHub using your UCLA email.

Yes I did.

2. Create a private repository biostat-m280-2018-winter and add Hua-Zhou and juhkim111 as your collaborators with write permission. Top directories of the repository should be hw1, hw2, … Maintain two branches master and develop. The develop branch will be your main playground, the place where you develop solution (code) to homework problems and write up report. The master branch will be your presentation area. Submit your homework files (R markdown file Rmd, html file converted from R markdown, all code and data sets to reproduce results) in master branch.

Yes I did.

3. Top directories of the repository should be hw1, hw2, … Maintain two branches master and develop. The develop branch will be your main playground, the place where you develop solution (code) to homework problems and write up report. The master branch will be your presentation area. Submit your homework files (R markdown file Rmd, html file converted from R markdown, all code and data sets to reproduce results) in master branch.

Yes I did.

##Q2. Linux Shell Commands

Use Linux shell commands to answer following questions.

1. How many persons are in the data set (statisticians call this n)? How many SNPs are in the data set (statisticians call this p)?

There are n = 959 people in the sataset. 

```{bash, echo= FALSE, eval = FALSE}
wc -l /home/m280-data/hw1/merge-geno.fam
```

There are 8348674 SNP's in the dataset.

```{bash, echo= FALSE, eval = FALSE}
wc -l /home/m280-data/hw1/merge-geno.bim
```


2. Which chromosomes does this data set contain? How many SNPs are in each chromosome?

This dataset contains chromosomes 1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21.


```{bash, echo= FALSE, eval = FALSE}
awk -F " " '!_[$1]++' /home/m280-data/hw1/merge-geno.bim | awk '{print $1}'
```

Below are the count of SNP's in each chromosome.
Chromosome Count:
Chromosome 1 has 1309299 SNP's
Chromosome 3 has 1215399 SNP's
Chromosome 5 has 1090185 SNP's
Chromosome 7 has 980944 SNP's
Chromosome 9 has 732013 SNP's
Chromosome 11 has 815860 SNP's
Chromosome 13 has 602809 SNP's
Chromosome 15 has 491208 SNP's
Chromosome 17 has 477990 SNP's
Chromosome 19 has 393615 SNP's
Chromosome 21 has 239352 SNP's

```{bash, echo= FALSE, eval = FALSE}
awk '{h[$1]++} END { for(k in h) print k, h[k] }' /home/m280-data/hw1/merge-geno.bim | sort -n

```

3. MAP4 (microtubule-associated protein 4) is a gene on chromosome 3 spanning positions 47,892,180 bp – 48,130,769 bp. How many SNPs are located within MAP4 gene?

There are 894 SNP's located within the MAP4 gene.

```{bash, echo= FALSE, eval = FALSE}
awk '$1 ==3  && $4 >= 47892180 && $4 <= 48130769 {count++} END {print count}' /home/m280-data/hw1/merge-geno.bim
```

4. Write a Linux shell command to convert merge-geno.fam to Mendel pedigree file. Since twin status is not available in plink format, we put nothing for that field. Also Mendel limits Person ID to have length less than or equal to 8 characters, so we have to strip the string T2DG from the IDs. 

```{bash, echo= FALSE, eval = FALSE}
#First we make a copy of the fam file to grant us permission to make edits!
cp /home/m280-data/hw1/merge-geno.fam newpedfile

#Then we will strip the string T2DG from the IDs.
sed -i 's/T2DG//' newpedfile
sed -i 's/T2DG//' newpedfile
sed -i 's/T2DG//' newpedfile

#make comma separated 
#make sex M F and save it to out.ped
awk '{OFS = ","; if ($5 == "1") {$5 = "M";} else if ($5 == "2") {$5 = "F";} else {$5 = "0";} print;}' newpedfile > out.ped 

#Then we make the last column of 0's into missing values

#this makes too many missing values...
#awk '{print $1, $2, $3, $4, $5, ""}' out.ped > out2.ped
#awk '{OFS = "\t"; if($3=="0") {$3="";} }' newpedfile


```

##Q3. R Batch Run
